includes/class-monk-links.php

```diff
@@ -0,0 +1,523 @@
+				return $form;
+			} else {
+				$form = str_replace( '</form>', '<input type="hidden" name="lang" value="' . esc_attr( $language ) . '" /></form>', $form );
+			}
+		}
+
+		return $form;
+	}
+
+	/**
+	 * Changes the language inside the given url
+	 *
+	 * @param  string $url provided url.
+	 * @param  string $lang the correct language to use.
+	 * @return $url
+	 */
+	public function monk_change_language_url( $url, $lang ) {
+		if ( empty( $lang ) ) {
+			return $url;
+		} else {
+
+			$active_languages = $this->monk_get_active_languages();
+
+			if ( $this->monk_using_permalinks() ) {
+
+				if ( ! empty( $active_languages ) ) {
+
+					$slug    = $lang . '/';
+					$pattern = str_replace( '/', '\/', $this->site_home . '/' . $this->site_root );
+					$pattern = '#' . $pattern . '(' . implode( '|', $active_languages ) . ')(\/|$)#';
+					$url     = preg_replace( $pattern, $this->site_home . '/' . $this->site_root, $url );
+
+					return str_replace( $this->site_home . '/' . $this->site_root, $this->site_home . '/' . $this->site_root . $slug, $url );
+				}
+			} else {
+				$url = remove_query_arg( 'lang', $url );
+				$url = ( empty( $lang ) ) ? $url : add_query_arg( 'lang', $lang, $url );
+				return $url;
+			}
+		}
+	}
+
+	/**
+	 * Redirects the incoming url when a wrong ( or lacking ) language is detected
+	 * preventing duplicated content for SEO performance
+	 *
+	 * @since  0.2.0
+	 * @return $redirect_url
+	 */
+	public function monk_need_canonical_redirect() {
+		global $wp_query, $post, $is_IIS;
+
+		// we do not want to redirect in these cases.
+		if ( is_search() || is_admin() || is_robots() || is_preview() || is_trackback() || ( $is_IIS && ! iis7_supports_permalinks() ) ) {
+			return;
+		}
+
+		if ( is_attachment() && filter_input( INPUT_GET, 'attachment_id' ) ) {
+			return;
+		}
+
+		if ( filter_input( INPUT_GET, 'wp_customize' ) || filter_input( INPUT_GET, 'customized' ) ) {
+			return;
+		}
+
+		// first get the correct url and scheme.
+		$requested_url = ( is_ssl() ? 'https://' : 'http://' ) . $_SERVER['HTTP_HOST'] . $_SERVER['REQUEST_URI'];
+
+		/**
+		 * Then decide what type of content is being requested
+		 * prior to get its language
+		 */
+		if ( is_single() || is_page() ) {
+			if ( isset( $post->ID ) ) {
+				$language = get_post_meta( $post->ID, '_monk_post_language', true );
+			}
+		} elseif ( is_category() || is_tax() || is_tag() ) {
+
+			$obj 	  = $wp_query->get_queried_object();
+			$language = get_term_meta( $obj->ID, '_monk_term_language', true );
+
+		} elseif ( $wp_query->is_posts_page ) {
+
+			$obj 	  = $wp_query->get_queried_object();
+			$language = get_post_meta( $obj->ID, '_monk_post_language', true );
+
+		} elseif ( is_404() && ! empty( $wp_query->query['page_id'] ) && $id = get_query_var( 'page_id' ) ) {
+
+			$language = get_post_meta( $id, '_monk_post_language', true );
+		}
+
+		if ( empty( $language ) ) {
+
+			$language     = get_option( 'monk_default_language', false );
```

Comment:
In the case we do not get any of the above listed content, the language is the site language, to avoid breaking the links with no redirection needed